src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1013 @@
+
+    // Stack maintaining whether or not the node we're serializing inherits from the underlying BSON
+    // data base class required for classes that use bsoncxx view types.
+    std::stack<bool> _curNodeInheritsUnderlyingBSONDataBase;
+
+    // Boolean value that tracks whether or not the current root BSON document or array is being
+    // treated as a root element.
+    bool _objAsRootElement;
+
+};  // BSONOutputArchive
+
+class BSONInputArchive : public InputArchive<BSONInputArchive> {
+    enum class InputNodeType { InObject, InEmbeddedObject, InEmbeddedArray, InRootElement };
+
+   public:
+    /**
+    * Construct a BSONInputArchive from an input stream of BSON data.
+    *
+    * @param stream
+    *    The stream from which to read BSON data.
+    */
+    BSONInputArchive(std::istream& stream)
+        : InputArchive<BSONInputArchive>(this),
+          _nextName(nullptr),
+          _readStream(stream),
+          _readFirstDoc(false) {
+    }
+
+   private:
+    /**
+     * Reads the next BSON document from the istream. This should be called whenever
+     * we are starting to load in a root element or root node.
+     */
+    void readNextDoc() {
+        // Determine the size of the BSON document in bytes.
+        // TODO: Only works on little endian.
+        int32_t docsize;
+        char docsize_buf[sizeof(docsize)];
+        _readStream.read(docsize_buf, sizeof(docsize));
+        std::memcpy(&docsize, docsize_buf, sizeof(docsize));
+
+        // Throw an exception if the end of the stream is prematurely reached.
+        if (_readStream.eof() || !_readStream || docsize < 5) {
+            throw cereal::Exception("No more data in BSONInputArchive stream.\n");
+        }
+
+        // Create a shared buffer to store the BSON document.
+        _curBsonData =
+            std::shared_ptr<uint8_t>{new uint8_t[docsize], [](uint8_t* p) { delete[] p; }};
+
+        // Read the BSON data from the stream into the buffer.
+        std::memcpy(_curBsonData.get(), docsize_buf, sizeof(docsize));
+        _readStream.read(reinterpret_cast<char*>(_curBsonData.get() + sizeof(docsize)),
+                         docsize - sizeof(docsize));
+
+        // Make sure there were no errors reading the BSON data.
+        if (_readStream.eof() || !_readStream) {
+            throw cereal::Exception("No more data in BSONInputArchive stream.\n");
+        }
+
+        // Store the BSON data of the document in a view that we can access.
+        _curBsonDoc = bsoncxx::document::view{_curBsonData.get(), static_cast<size_t>(docsize)};
+        _curBsonDataSize = docsize;
+
+        // Specify that we've read a document.
+        _readFirstDoc = true;
+    }
+
+    /**
+     * Searches for the next BSON element to be retrieved and loaded.
+     *
+     * @return the element with _nextName as a key, or if the current node is an
+     *         array, the next element in that array.
+     */
+    inline bsoncxx::types::value search() {
```

Comment:
Cereal works by linearly iterating through the fields we have to populate (and recursively iterating through fields in subdocuments), so this would be pretty tricky.

We could theoretically create a `std::map` hash table when we read in the document in `readNextDoc()` and use that in `search()` instead of the `bsoncxx::document::view`'s `[ ]` overload.

I think it would be the most feasible way to do this, because trying to fill in the fields if they exist as we read through the BSON just isn't possible with the way cereal works. What do you think?